File: application/smartgarden_flutter_socket/lib/src/models/pump_state.dart
class PumpStateModel {
  final int nodeId;
  final String state; // "ON" | "OFF"
  final String source; // "manual" | "auto"
  final DateTime? expiresAt;
  final bool manualLock;

  PumpStateModel({
    required this.nodeId,
    required this.state,
    required this.source,
    this.expiresAt,
    this.manualLock = false,
  });

  factory PumpStateModel.fromJson(Map<String, dynamic> j) => PumpStateModel(
        nodeId: j["node_id"],
        state: j["state"],
        source: j["source"] ?? "manual",
        expiresAt:
            j["expiresAt"] != null ? DateTime.parse(j["expiresAt"]) : null,
        manualLock: j["manualLock"] == true,
      );
}

File: server/src/models/PumpState.js
import mongoose from "mongoose";

const PumpStateSchema = new mongoose.Schema(
    {
        node_id: { type: Number, index: true, required: true, unique: true },
        state: { type: String, enum: ["ON", "OFF"], default: "OFF" },
        source: { type: String, enum: ["manual", "auto"], default: "manual" },
        // Khi source=auto và có thời hạn
        expiresAt: { type: Date },
        // Manual lock: khi true thì automation KHÔNG được override cho tới khi user tắt manual
        manualLock: { type: Boolean, default: false },
        updatedAt: { type: Date, default: Date.now },
    },
    { timestamps: true }
);

export default mongoose.model("PumpState", PumpStateSchema);
